---
title: "TFM-code"
author: "Lorena Ponce Ruiz"
date: '`r format(Sys.Date(),"%e, %B, %Y")`'
lang: es
output: 
  html_document:
    keep_md: false
    toc: true
    toc_float: true
    toc_depth: 4
    code_download: true
---

```{r, eval = FALSE}
setwd("E:/Escritorio/Epigenetics & cancer")
```

```{r isntall_Rpackages, include=FALSE, eval = FALSE}
install.packages("devtools")
BiocManager::install("maftools", version = "3.8")
BiocManager::install("knitr", version = "3.8")
BiocManager::install("curatedTCGAData", version = "3.8")
BiocManager::install("MultiAssayExperiment", version = "3.8")
BiocManager::install("RaggedExperiment", version = "3.8")
BiocManager::install("TCGAbiolinks", version = "3.8")
BiocManager::install("ExperimentHub", version = "3.8")
BiocManager::install("stringr", version = "3.8")
BiocManager::install("IlluminaHumanMethylation450kanno.ilmn12.hg19", version = "3.8")
BiocManager::install("limma", version = "3.8")
BiocManager::install("ABC.RAP", version = "3.8")
```

```{r Rpackages, include=FALSE}
library(ggplot2)
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(curatedTCGAData))
#devtools::install_github('BioinformaticsFMRP/TCGAbiolinks')
suppressPackageStartupMessages(library(TCGAbiolinks))
suppressPackageStartupMessages(library(MultiAssayExperiment))
suppressPackageStartupMessages(library(RaggedExperiment))
#suppressPackageStartupMessages(library(ExperimentHub))
#suppressPackageStartupMessages(library(devtools))
## library(biomaRt)
suppressPackageStartupMessages(library(IlluminaHumanMethylation450kanno.ilmn12.hg19))
suppressPackageStartupMessages(library(stringr))
suppressPackageStartupMessages(library("maftools"))
library(ComplexHeatmap)
require(limma)
```

```{r constants, cache=FALSE}
NUM_PATIENTS <- 50
```

```{r commonBarcodes_func, cache=FALSE}
##' it samples the patients, if the cohort has more than n patients with matched primary tumor and normal
##' @title 
##' @param project the name of the TCGA project
##' @param query a GDC query result
##' @param n the number of patients to sample
##' @return a named list with all the barcodes for primary tumors and normals, max size 2n
##' @author Lorena Ponce

commonBarcodes <- function(project, query, n){
  samplesMetExp <- matchedMetExp(project, legacy = TRUE)
  results <- getResults(query, cols = c("cases", "tissue.definition"))
  ids_tumor <- results$cases[which(results$tissue.definition == "Primary solid Tumor")]
  ids_normal <- results$cases[which(results$tissue.definition == "Solid Tissue Normal")]
  
  common <- intersect(strtrim(ids_tumor, 12), strtrim(ids_normal, 12))
  commonMetExp <- intersect(strtrim(common, 12), strtrim(samplesMetExp, 12))
  
  ## sampling to n patients, if getting more than those
  if (n < length(commonMetExp)) {
    set.seed(1)
    commonMetExp <- sample(commonMetExp, n)
  }
  
  named_tumors <- ids_tumor
  names(named_tumors) <- strtrim(ids_tumor, 12)
  named_normals <- ids_normal
  names(named_normals) <- strtrim(ids_normal, 12)
  barcodes <- as.list(c(named_tumors[commonMetExp], named_normals[commonMetExp]))
  return(barcodes)
}
```


```{r annotation}
ann450k = getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
```


# Download methylation 450K dataset for: 

## COAD (colon cancer)
```{r met.coad}
coad <- "TCGA-COAD"

query.met.coad <- GDCquery(project = coad,
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450",
                          sample.type = c("Primary solid Tumor", "Solid Tissue Normal"))

barcodes.coad <- commonBarcodes(coad, query.met.coad, NUM_PATIENTS)

query.met.coad2 <- GDCquery(project = coad,
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450",
                          sample.type = c("Primary solid Tumor", "Solid Tissue Normal"),
                          barcode = barcodes.coad)

#GDCdownload(query.met.coad2)

met.coad.450 <- GDCprepare(query = query.met.coad2,
                         save = FALSE,
                         save.filename = "coadDNAmet450k.rda" ,
                         summarizedExperiment = TRUE)
```

## BRCA (breast cancer)
```{r met.brca}
brca <- "TCGA-BRCA"

query.met.brca <- GDCquery(project = brca,
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450",
                          sample.type = c("Primary solid Tumor", "Solid Tissue Normal"))

barcodes.brca <- commonBarcodes(brca, query.met.brca, NUM_PATIENTS)

query.met.brca2 <- GDCquery(project = brca,
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450",
                          sample.type = c("Primary solid Tumor", "Solid Tissue Normal"),
                          barcode = barcodes.brca)

#GDCdownload(query.met.brca2)

met.brca.450 <- GDCprepare(query = query.met.brca2,
                         save = FALSE,
                         save.filename = "brcaDNAmet450k.rda" ,
                         summarizedExperiment = TRUE)
```

## LUAD (lung cancer)
```{r met.luad}
luad <- "TCGA-LUAD"

query.met.luad <- GDCquery(project = luad,
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450",
                          sample.type = c("Primary solid Tumor", "Solid Tissue Normal"))

barcodes.luad <- commonBarcodes(luad, query.met.luad, NUM_PATIENTS)

query.met.luad2 <- GDCquery(project = luad,
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450",
                          sample.type = c("Primary solid Tumor", "Solid Tissue Normal"),
                          barcode = barcodes.luad)

#GDCdownload(query.met.luad2)

met.luad.450 <- GDCprepare(query = query.met.luad2,
                         save = FALSE,
                         save.filename = "luadDNAmet450k.rda" ,
                         summarizedExperiment = TRUE)

```

## PRAD (prostate cancer)
```{r met.prad}
prad <- "TCGA-PRAD"

query.met.prad <- GDCquery(project = prad,
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450",
                          sample.type = c("Primary solid Tumor", "Solid Tissue Normal"))

barcodes.prad <- commonBarcodes(prad, query.met.prad, NUM_PATIENTS)

query.met.prad2 <- GDCquery(project = prad,
                          legacy = TRUE,
                          data.category = "DNA methylation",
                          platform = "Illumina Human Methylation 450",
                          sample.type = c("Primary solid Tumor", "Solid Tissue Normal"),
                          barcode = barcodes.prad)

#GDCdownload(query.met.prad2)

met.prad.450 <- GDCprepare(query = query.met.prad2,
                         save = FALSE,
                         save.filename = "pradDNAmet450k.rda" ,
                         summarizedExperiment = TRUE)
```


# Download gene expression (normalized) datasets for:

## COAD
```{r exp.coad}
query.exp.coad <- GDCquery(project = coad, 
                      legacy = TRUE,
                      data.category = "Gene expression",
                      data.type = "Gene expression quantification",
                      platform = "Illumina HiSeq", 
                      file.type = "normalized_results",
                      sample.type = c("Primary solid Tumor", "Solid Tissue Normal"),
                      barcode = strtrim(barcodes.coad, 12))

#GDCdownload(query.exp.coad)

exp.coad <- GDCprepare(query = query.exp.coad,
                      save = FALSE,
                      save.filename = "coadRNAseq.rda" ,
                      summarizedExperiment = TRUE)
```


## BRCA
```{r exp.brca}
query.exp.brca <- GDCquery(project = brca, 
                      legacy = TRUE,
                      data.category = "Gene expression",
                      data.type = "Gene expression quantification",
                      platform = "Illumina HiSeq", 
                      file.type = "normalized_results",
                      sample.type = c("Primary solid Tumor", "Solid Tissue Normal"), 
                      barcode = strtrim(barcodes.brca, 12))

## obtener el listado de barcodes no duplicados en exp.brca
dup <- query.exp.brca$results[[1]]$cases[duplicated(query.exp.brca$results[[1]]$cases)]
barcodes.no.dup <- query.exp.brca$results[[1]]$cases[!duplicated(query.exp.brca$results[[1]]$cases)]
idx <- which(barcodes.no.dup %in% dup) #listado con el ?ndice de los barcodes duplicados
barcodes.no.dup <- barcodes.no.dup[-idx]
any(dup %in% barcodes.no.dup)

query.exp.brca <- GDCquery(project = brca, 
                      legacy = TRUE,
                      data.category = "Gene expression",
                      data.type = "Gene expression quantification",
                      platform = "Illumina HiSeq", 
                      file.type = "normalized_results",
                      sample.type = c("Primary solid Tumor", "Solid Tissue Normal"), 
                      barcode = barcodes.no.dup)

#GDCdownload(query.exp.brca)

exp.brca <- GDCprepare(query = query.exp.brca,
                      save = FALSE,
                      save.filename = "brcaRNAseq.rda" ,
                      summarizedExperiment = TRUE)
```


## LUAD
```{r exp.luad}
query.exp.luad <- GDCquery(project = luad, 
                      legacy = TRUE,
                      data.category = "Gene expression",
                      data.type = "Gene expression quantification",
                      platform = "Illumina HiSeq", 
                      file.type = "normalized_results",
                      sample.type = c("Primary solid Tumor", "Solid Tissue Normal"),
                      barcode = strtrim(barcodes.luad, 12))

GDCdownload(query.exp.luad)

exp.luad <- GDCprepare(query = query.exp.luad,
                      save = FALSE,
                      save.filename = "luadRNAseq.rda" ,
                      summarizedExperiment = TRUE)
```


## PRAD
```{r exp.prad}
query.exp.prad <- GDCquery(project = prad, 
                      legacy = TRUE,
                      data.category = "Gene expression",
                      data.type = "Gene expression quantification",
                      platform = "Illumina HiSeq", 
                      file.type = "normalized_results",
                      sample.type = c("Primary solid Tumor", "Solid Tissue Normal"),
                      barcode = strtrim(barcodes.prad, 12))

GDCdownload(query.exp.prad)

exp.prad <- GDCprepare(query = query.exp.prad,
                      save = FALSE,
                      save.filename = "pradRNAseq.rda",
                      summarizedExperiment = TRUE)
```


# Download mutations dataset with MAF format and CNV dataset:

```{r mutation.download}
## for cOAD
query.mut.coad  <- GDCquery(project = coad, 
                            data.category = "Simple nucleotide variation", 
                            data.type = "Simple somatic mutation",
                            sample.type = "Primary solid Tumor",
                            legacy = TRUE)

#GDCdownload(query.mut.coad)

mut.coad <- GDCprepare(query.mut.coad,
                      save = TRUE,
                      save.filename = "coadMut.rda")

#maf.coad <- curatedTCGAData(diseaseCode = "COAD",
#                        assays = "Mutation",
#                        dry.run = FALSE)

cna.coad <- curatedTCGAData(diseaseCode = "COAD",
                            assays = "*GISTIC*",
                            dry.run = FALSE)

## for BRCA
#query.mut.brca  <- GDCquery(project = brca, 
#                            data.category = "Simple nucleotide variation", 
#                            data.type = "Simple somatic mutation",
#                            sample.type = "Primary solid Tumor",
#                            file.type = "gsc_BRCA_pairs.aggregated.capture.tcga.uuid.automated.somatic.maf",
#                            legacy = TRUE)

#GDCdownload(query.mut.brca)
#maf.brca <- GDCprepare(query.mut.brca, save = TRUE, save.filename = "brcaMut.rda")

maf.brca <- curatedTCGAData(diseaseCode = "BRCA",
                        assays = "Mutation",
                        dry.run = FALSE)

cna.brca <- curatedTCGAData(diseaseCode = "BRCA",
                            assays = "*GISTIC*",
                            dry.run = FALSE)

## for LUAD
#query.mut.luad  <- GDCquery(project = luad, 
#                            data.category = "Simple nucleotide variation", 
#                            data.type = "Simple somatic mutation",
#                            #sample.type = "Primary solid Tumor",
#                            legacy = TRUE)
#GDCdownload(query.mut.luad)
#maf.luad <- GDCprepare(query.mut.luad,
#                      save = TRUE,
#                      save.filename = "luadMut.rda")

maf.luad <- curatedTCGAData(diseaseCode = "LUAD",
                        assays = "Mutation",
                        dry.run = FALSE)

cna.luad <- curatedTCGAData(diseaseCode = "LUAD",
                            assays = "*GISTIC*",
                            dry.run = FALSE)

## for PRAD
#query.mut.prad  <- GDCquery(project = prad, 
#                            data.category = "Simple nucleotide variation", 
#                            data.type = "Simple somatic mutation",
#                            sample.type = "Primary solid Tumor",
#                            legacy = TRUE)
#GDCdownload(query.mut.prad)
#maf.prad <- GDCprepare(query.mut.prad,
#                      save = TRUE,
#                      save.filename = "pradMut.rda")

maf.prad <- curatedTCGAData(diseaseCode = "PRAD",
                        assays = "Mutation",
                        dry.run = FALSE)

cna.prad <- curatedTCGAData(diseaseCode = "PRAD",
                            assays = "*GISTIC*",
                            dry.run = FALSE)
```


# Download clinical data from samples downloaded:

```{r clinical.download, eval=FALSE}
# NOT USED!!
## for COAD
query.clinic.coad <- GDCquery(project = coad, 
                  data.category = "Clinical",
                  data.type = "Clinical data",
                  legacy = TRUE,
                  file.type = "xml",
                  barcode = strtrim(barcodes.coad, 12)) 

GDCdownload(query.clinic.coad)

clinical.coad <- GDCprepare(query.clinic.coad,
                                      save = FALSE,
                                      save.filename = "brcaClinic.rda")


## for BRCA
query.clinic.brca <- GDCquery(project = brca, 
                  data.category = "Clinical",
                  data.type = "Clinical data",
                  legacy = TRUE,
                  file.type = "xml",
                  barcode = strtrim(barcodes.brca, 12)) 

GDCdownload(query.clinic.brca)

clinical.brca <- GDCprepare(query.clinic.brca,
                                      save = FALSE,
                                      save.filename = "brcaClinic.rda")

## for LUAD
query.clinic.luad <- GDCquery(project = luad, 
                  data.category = "Clinical",
                  data.type = "Clinical data",
                  legacy = TRUE,
                  file.type = "xml",
                  barcode = strtrim(barcodes.luad, 12)) 

GDCdownload(query.clinic.luad)

clinical.luad <- GDCprepare(query.clinic.luad,
                                      save = FALSE,
                                      save.filename = "luadClinic.rda")

## for PRAD
query.clinic.prad <- GDCquery(project = prad, 
                  data.category = "Clinical",
                  data.type = "Clinical data",
                  legacy = TRUE,
                  file.type = "xml",
                  barcode = strtrim(barcodes.prad, 12)) 

GDCdownload(query.clinic.prad)

clinical.prad <- GDCprepare(query.clinic.prad,
                                      save = FALSE,
                                      save.filename = "pradClinic.rda")
```


```{r clinical.download2}
clinic_coad <- GDCquery_clinic(coad, type = "clinical", save.csv = FALSE)
clinic_brca <- GDCquery_clinic(brca, type = "clinical", save.csv = FALSE)
clinic_prad <- GDCquery_clinic(prad, type = "clinical", save.csv = FALSE)
clinic_luad <- GDCquery_clinic(luad, type = "clinical", save.csv = FALSE)
```


Commented out as it raises

```
> luad.map <- sampleMap(luad)
Error in (function (classes, fdef, mtable)  :
  unable to find an inherited method for function ‘sampleMap’ for signature ‘"character"’


```

```{r, eval = FALSE}
## Check if the mutation and methylation barcodes match
## LUAD with curatedTCGAData
## luad.map<- sampleMap(luad)
length(which(strtrim(barcodes.luad, 15) %in% strtrim(luad.map$colname, 15)))
cna.luad.data <- sampleMap(cna.luad)
length(which(strtrim(barcodes.luad, 15) %in% strtrim(cna.luad.data$colname, 15)))

##COAD
#with TCGABiolinks
mutation.coad.data <- sampleMap(mutation.coad)
length(which(strtrim(barcodes.coad, 12) %in% strtrim(mutation.coad.data$colname, 12)))
length(which(strtrim(barcodes.coad, 15) %in% strtrim(maf.coad$Tumor_Sample_Barcode, 15)))
cna.coad.data <- sampleMap(cna.coad)
length(which(strtrim(barcodes.coad, 15) %in% strtrim(cna.coad.data$colname, 15)))

#with curatedTCGAData
coad.map <- sampleMap(coad)
which(strtrim(barcodes.coad, 12) %in% strtrim(coad.map$colname, 12))

##BRCA with TCGABiolinks
length(which(strtrim(barcodes.brca, 15) %in% strtrim(maf.brca$Tumor_Sample_Barcode, 15)))
brca.map <- sampleMap(brca)
length(which(strtrim(barcodes.brca, 15) %in% strtrim(brca.map$colname, 15)))
mutation. <- sampleMap(brca)
length(which(strtrim(barcodes.brca, 15) %in% strtrim(brca.map$colname, 15)))
cna.brca.data <- sampleMap(cna.brca)
length(which(strtrim(barcodes.brca, 15) %in% strtrim(cna.brca.data$colname, 15)))


##PRAD with curatedTCGAData
prad.map<- sampleMap(prad)
length(which(strtrim(barcodes.prad, 15) %in% strtrim(prad.map$colname, 15)))
cna.prad.data <- sampleMap(cna.prad)
length(which(strtrim(barcodes.prad, 15) %in% strtrim(cna.prad.data$colname, 15)))
```


# PIPELINE PARA EL AN?LSIS DEL METILOMA 

Con los datos descargados con TCGABiolinks (level = 3) se obtienen los valores beta de metilaci?n, los cuales est?n escalados de 0.0 (sin metilar) a 1.0 (totalmente metilados).

## BÚSQUEDA DE REGIONES CON UN PERFIL DE METILACI?N ABERRANTE

### PREVISUALIZACI?N DE LAS BASES DE DATOS

```{r dim.met}
# VISUALIZACI?N DE LOS DATOS 

dim(met.coad.450)
dim(met.brca.450)
dim(met.prad.450)
dim(met.luad.450)
```

Los datos de la plataforma de metilaci?n 450k, llamada as? por el n?mero de probes que analiza, presenta 3 tipos de probes: cg (CpG loci), ch (non-CpG loci) and rs (SNP assay). Como se puede comprobar todas las bases de datos disponen del mismo n?mero de filas que se corresponden con las probes descargadas y el n?mero de columnas se corresponden con el n?mero de muestras.


### PREPROCESAMIENTO DE LOS DATOS
A continuaci?n se preprocesan los datos obtenidos de metilaci?n:

```{r clean.met}
# PREPROCESAMIENTO DE LOS DATOS

## eliminamos los valores faltantes NA
met.coad <- subset(met.coad.450,subset = (rowSums(is.na(assay(met.coad.450))) == 0))
dim(met.coad)
met.brca <- subset(met.brca.450,subset = (rowSums(is.na(assay(met.brca.450))) == 0))
dim(met.brca)
met.prad <- subset(met.prad.450,subset = (rowSums(is.na(assay(met.prad.450))) == 0))
dim(met.prad)
met.luad <- subset(met.luad.450,subset = (rowSums(is.na(assay(met.luad.450))) == 0))
dim(met.luad)

## eliminamos los datos procedentes de los cromosomas sexuales (chrX y chrY) y aquellos correspondientes a las probes rs, que han de eliminarse de acuerdo a los manuales de Illumina 
met.coad <- subset(met.coad,subset = !as.character(seqnames(met.coad)) %in% c("chrNA","chrX","chrY"))
dim(met.coad)
met.brca <- subset(met.brca,subset = !as.character(seqnames(met.brca)) %in% c("chrNA","chrX","chrY"))
dim(met.brca)
met.prad <- subset(met.prad,subset = !as.character(seqnames(met.prad)) %in% c("chrNA","chrX","chrY"))
dim(met.prad)
met.luad <- subset(met.luad,subset = !as.character(seqnames(met.luad)) %in% c("chrNA","chrX","chrY"))
dim(met.luad)
```


### PREVISUALIZACIÓN DE LOS DATOS


```{r densityPlot.betaVal}
par(mfrow = c(2,2))
densityPlot(assay(met.coad), sampGroups =colData(met.coad)$definition, legend = TRUE, main = "COAD beta-values")
densityPlot(assay(met.brca), sampGroups =colData(met.brca)$definition, legend = TRUE, main = "BRCA beta-values")
densityPlot(assay(met.luad), sampGroups =colData(met.luad)$definition, legend = TRUE, main = "LUAD beta-values")
densityPlot(assay(met.prad), sampGroups =colData(met.prad)$definition, legend = TRUE, main = "PRAD beta-values")
```



```{r qplot.met.coad, eval = FALSE}
# DISTRIBUCIÓN EN COAD

qplot(assay(met.coad)[ , grepl( "01A" , colnames( assay(met.coad) ) ) ] , fill=I(rgb(0.1,0.2,0.4,0.6)), main = "Tejido canceroso COAD", xlab = "Beta-values", ylab = "recuento", bins = 50)
qplot(assay(met.coad)[ , grepl( "11A" , colnames( assay(met.coad) ) ) ] , fill=I(rgb(0.1,0.2,0.4,0.6)), main = "Tejido normal COAD", xlab = "Beta-values", ylab= "recuento", bins = 50 )
```

En las distribución de los beta values de COAD se aprecian muchos valores cercanos a 0.50, posible razón por la que al calcular la media entre ambos tipos de muestras y estudiar la significación no se aprecien diferencias significativas el grado de metilación. 

```{r qplot.met.brca, eval = FALSE}
# DISTRIBUCIÓN EN BRCA

qplot(assay(met.brca)[ , grepl( "01A" , colnames( assay(met.brca) ) ) ] , fill=I(rgb(0.72, 0.201, 0.176,0.6)), main = "Tejido canceroso BRCA", xlab = "Beta-values", ylab = "recuento", bins = 50, ylim = c(0, 50000))
qplot(assay(met.brca)[ , grepl( "11A" , colnames( assay(met.brca) ) ) ] , fill=I(rgb(0.1,0.2,0.4,0.6)), main = "Tejido normal BRCA", xlab = "Beta-values", ylab= "recuento", bins = 50, ylim = c(0, 50000) )
```

Comparando los valores beta gráficamente de ambos tipos de tejido, apenas hay diferencias entre tejido normal y canceroso en BRCA.

```{r qplot.met.prad, eval = FALSE}
# DISTRIBUCIÓN EN PRAD

qplot(assay(met.prad)[ , grepl( "01A" , colnames( assay(met.prad) ) ) ] , fill=I(rgb(0.1,0.2,0.4,0.6)), main = "Tejido canceroso PRAD", xlab = "Beta-values", ylab = "recuento", bins = 50)
qplot(assay(met.prad)[ , grepl( "11A" , colnames( assay(met.prad) ) ) ] , fill=I(rgb(0.1,0.2,0.4,0.6)), main = "Tejido normal PRAD", xlab = "Beta-values", ylab= "recuento", bins = 50 )
```

En PRAD parece que aumenta los valores beta en las muestras de tejido canceroso en comparación con las muestras de tejido normal, por lo que podría presentar un marcado perfil de hipermetilación.



```{r qplot.met.luad, eval = FALSE}
# DISTRIBUCIÓN EN LUAD

qplot(assay(met.luad)[ , grepl( "01A" , colnames( assay(met.luad) ) ) ] , fill=I(rgb(0.1,0.2,0.4,0.6)), main = "Tejido canceroso LUAD", xlab = "Beta-values", ylab = "recuento", bins = 50)
qplot(assay(met.luad)[ , grepl( "11A" , colnames( assay(met.luad) ) ) ] , fill=I(rgb(0.1,0.2,0.4,0.6)), main = "Tejido normal LUAD", xlab = "Beta-values", ylab= "recuento", bins = 50 )
```

LUAD es la base de datos que contiene menos muestras (N= 28 por tipo de tejido). A pesar de que se aprecia un cambio en la distribución de los valores de beta, en los análsis posteriores no se aprecia diferencias significativas. A pesar de que los cambios observados, a priori, no parecen significativos, hay una cierta migración de valores cercanos a 1 (metilados) en tejido normal y valores más intermedios en tejido canceroso. Podría indicar un perfil de hipometilación en este tipo de cáncer. 



### DIFERENCIAS ENTRE LAS MEDIAS DE METILACI?N DE MUSTRAS DE TEJIDO NORMAL Y TUMORAL
Seguidamente se realizan un c?lculo descriptivo previo, en el que apreciar si hay diferencias significativas entre la media de metilaci?n etre muestras de tejido normal y tumoral. Para ello calcula la media de metilaci?n (la media entre las betas) de cada una de las muestras, a partir de ah? se realizan los diagramas de cajas con estos datos.

```{r meanNT.coad}
# VISUALIZACI?N DE MEDIAS ENTRE BETAS DE MUESTRAS NORMALES Y TUMORALES PARA COAD

TCGAvisualize_meanMethylation(met.coad,
                              groupCol = "definition",
                              group.legend = "Groups",
                              print.pvalue = TRUE,
                              filename = 'methylationMean_COAD_norm-tum.pdf')
```

Con un p-valor igual a 0.196 no hay diferencias significativas entre el grado de metilaci?n en las muestras normales y tumorales para el tipo de c?ncer de colon. Al observar el m?nimo y el m?ximo en ambos grupos se aprecia que las muestras tumorales presentan mayores perfiles de hipo e hipermetilaci?n. 


Confirma lo apreciado con los diagramas de cajas empleados con el paquete TCGAbiolinks, apenas hay diferencias significativas entre el conjunto de probes de COAD.

```{r meanNT.brca}
# VISUALIZACI?N DE MEDIAS ENTRE BETAS DE MUESTRAS NORMALES Y TUMORALES PARA BRCA

TCGAvisualize_meanMethylation(met.brca,
                              groupCol = "definition",
                              group.legend = "Groups",
                              print.pvalue = TRUE,
                              filename = 'methylationMean_BRCA_norm-tum.pdf')
```

Al igual que ocurre para COAD, no hay diferencias significativas al comparar el grado de metilaci?n entre muestras de tejido normal y canceroso.


```{r meanNT.prad}
# VISUALIZACI?N DE MEDIAS ENTRE BETAS DE MUESTRAS NORMALES Y TUMORALES PARA PRAD

TCGAvisualize_meanMethylation(met.prad,
                              groupCol = "definition",
                              group.legend = "Groups",
                              print.pvalue = TRUE,
                              filename = 'methylationMean_PRAD_norm-tum.pdf')
```

En PRAD **s? que hay diferencias significativas**.


```{r meanNT.luad}
# VISUALIZACI?N DE MEDIAS ENTRE BETAS DE MUESTRAS NORMALES Y TUMORALES PARA LUAD

TCGAvisualize_meanMethylation(met.luad,
                              groupCol = "definition",
                              group.legend = "Groups",
                              print.pvalue = TRUE,
                              filename = 'methylationMean_LUAD_norm-tum.pdf')
```

En LUAD no solo no hay diferencias significativas, sino que es la cohorte en la que se aprecia una menor diferencia entre ambos tipos de muestras.


### REGIONES DIFERENCIALMENTE METILADAS (DMR) CON TCGABiolinks
A continuaci?n se obtendr?n las regiones diferencialmente metiladas en ambos grupos (normal y tumoral), para ello el paquete TCGABiolinks dispone de una funci?n `TCGAanalyze_DMR`. Esta funci?n calcula en primer lugar la diferencia entre la media de metilizaci?n (media de los valores beta) de cada grupo para cada probe. Seguidamente, calcula la expresi?n diferencial empleando el test de Wilcoxon ajustado por el m?todo de Benjamini-Hochberg. Se especifican adem?s el valor umbral del p-valor ajustado y la diferencia absoluta m?nima entre los valores beta.

1. COAD

```{r dmr.coad.tcga}
# CALCULO DMR (Differentially Methylation Region) para COAD

met.coad <- TCGAanalyze_DMR(met.coad,
                groupCol = "definition", 
                group1 = "Primary solid Tumor",
                group2= "Solid Tissue Normal", 
                p.cut = 0.01,
                diffmean.cut = 0.25,
                legend = "State",
                plot.filename = "metvolcano_COAD.png",
                cores = 1)
```


```{r dmr.sig.coad}
dmr_coad <- read.csv("DMR_results_definition_Primary.solid.Tumor_Solid.Tissue.Normal_pcut_0.01_meancut_0.25.csv")
#"DMR_results_COAD_definition_Primary.solid.Tumor_Solid.Tissue.Normal_pcut_0.01_meancut_0.25.csv"

## exploraci?n de los datos de metilaci?n diferencial entre muestras normales y tumorales
dim(dmr_coad)
length(which(dmr_coad$status.Solid.Tissue.Normal.Primary.solid.Tumor != "Not Significant"))

cols <- c("status.Primary.solid.Tumor.Solid.Tissue.Normal", "status.Solid.Tissue.Normal.Primary.solid.Tumor")
summary(dmr_coad[cols])

dmr_coad_sel <- dmr_coad[dmr_coad$status.Solid.Tissue.Normal.Primary.solid.Tumor %in% c("Hypermethylated in Primary solid Tumor", "Hypomethylated in Primary solid Tumor"), -c(5:8)]

dmr_coad_sel$status.Solid.Tissue.Normal.Primary.solid.Tumor <- factor(dmr_coad_sel$status.Solid.Tissue.Normal.Primary.solid.Tumor, levels = c("Hypermethylated in Primary solid Tumor", "Hypomethylated in Primary solid Tumor"), labels = c('Hypermethylated', 'Hypomethylated'))
```


2. BRCA

```{r dmr.brca.tcga}
# CALCULO DMR (Differentially Methylation Region) para BRCA

met.brca <-TCGAanalyze_DMR(met.brca,
                groupCol = "definition", 
                group1 = "Primary solid Tumor",
                group2= "Solid Tissue Normal", 
                p.cut = 0.01,
                diffmean.cut = 0.25,
                legend = "State",
                plot.filename = "metvolcano_BRCA.png",
                cores = 1)
```

```{r dmr.sig.brca}
dmr_brca <- read.csv("DMR_results_definition_Primary.solid.Tumor_Solid.Tissue.Normal_pcut_0.01_meancut_0.25.csv")
#"DMR_results_BRCA_definition_Primary.solid.Tumor_Solid.Tissue.Normal_pcut_0.01_meancut_0.25.csv"

## exploraci?n de los datos de metilaci?n diferencial entre muestras normales y tumorales
dim(dmr_brca)
length(which(dmr_brca$status.Solid.Tissue.Normal.Primary.solid.Tumor != "Not Significant"))

summary(dmr_brca[cols])

dmr_brca_sel <- dmr_brca[dmr_coad$status.Solid.Tissue.Normal.Primary.solid.Tumor %in% c("Hypermethylated in Primary solid Tumor", "Hypomethylated in Primary solid Tumor"), -c(5:8)]
```
Parece que COAD presenta mayor grado de metilaci?n que BRCA, y en ambos casos es mayor el n?mero de regiones hipermetiladas que hipometiladas.


3. PRAD

```{r dmr.prad.tcga}
# CALCULO DMR (Differentially Methylation Region) para BRCA

met.prad <-TCGAanalyze_DMR(met.prad,
                groupCol = "definition", 
                group1 = "Primary solid Tumor",
                group2= "Solid Tissue Normal", 
                p.cut = 0.01,
                diffmean.cut = 0.25,
                legend = "State",
                plot.filename = "metvolcano_PRAD.png",
                cores = 1)
```

```{r dmr.sig.prad}
dmr_prad <- read.csv("DMR_results_definition_Primary.solid.Tumor_Solid.Tissue.Normal_pcut_0.01_meancut_0.25.csv")
#"DMR_results_PRAD_definition_Primary.solid.Tumor_Solid.Tissue.Normal_pcut_0.01_meancut_0.25.csv"

## exploraci?n de los datos de metilaci?n diferencial entre muestras normales y tumorales
dim(dmr_prad)
length(which(dmr_prad$status.Solid.Tissue.Normal.Primary.solid.Tumor != "Not Significant"))

summary(dmr_prad[cols])

dmr_prad_sel <- dmr_prad[dmr_prad$status.Solid.Tissue.Normal.Primary.solid.Tumor %in% c("Hypermethylated in Primary solid Tumor", "Hypomethylated in Primary solid Tumor"), -c(5:8)]
```


4. LUAD

```{r dmr.luad.tcga}
# CALCULO DMR (Differentially Methylation Region) para BRCA

met.luad <-TCGAanalyze_DMR(met.luad,
                groupCol = "definition", 
                group1 = "Primary solid Tumor",
                group2= "Solid Tissue Normal", 
                p.cut = 0.01,
                diffmean.cut = 0.25,
                legend = "State",
                plot.filename = "metvolcano_LUAD.png",
                cores = 1)
```

```{r dmr.sig.luad}
dmr_luad <- read.csv("DMR_results_definition_Primary.solid.Tumor_Solid.Tissue.Normal_pcut_0.01_meancut_0.25.csv")
#"DMR_results_LUAD_definition_Primary.solid.Tumor_Solid.Tissue.Normal_pcut_0.01_meancut_0.25.csv"

## exploraci?n de los datos de metilaci?n diferencial entre muestras normales y tumorales
dim(dmr_luad)
length(which(dmr_luad$status.Primary.solid.Tumor.Solid.Tissue.Normal != "Not Significant"))

summary(dmr_luad[cols])

dmr_luad_sel <- dmr_luad[dmr_prad$status.Solid.Tissue.Normal.Primary.solid.Tumor %in% c("Hypermethylated in Primary solid Tumor", "Hypomethylated in Primary solid Tumor"), -c(5:8)]
```


## INTEGRAR EPIGEN?MICA CON TRANSCRIPT?MICA

A continuaci?n se integran los datos obtenidos del metiloma con los datos de expresi?n, para comprobar qu? genes alteran su expresi?n debdido al perfil de metilaci?n (posteriormente buscar si alguno de estos genes presentan mutaciones caracter?sticas en alguna de las cohortes, especialmente de los genes candidatos).


1. COAD

```{r starburst.coad}
#dataNorm <-  TCGAanalyze_Normalization(tabDF = exp.coad, geneInfo =  geneInfo, method = 'gcContent') 
#dataNorm2 <-  TCGAanalyze_Normalization(tabDF = exp.coad, geneInfo =  geneInfo, method = 'geneLength')
#dataFilt <- TCGAanalyze_Filtering(tabDF = dataNorm2, method = "quantile", qnt.cut =  0.25)
#par(mfrow=c(2,2))
#boxplot(assay(exp.coad), main = "Datos sin normalizar")
#boxplot(dataNorm, main = "M?todo GC content")
#boxplot(dataNorm2, main = " M?todo geneLength")
#boxplot(dataFilt, main = "Datos normalizads \n + filtrados")


COAD.expMatrix <- assay(exp.coad)
samplesNT <- which(exp.coad$definition == 'Solid Tissue Normal')
samplesTP <- which(exp.coad$definition == 'Primary solid Tumor')

dea.coad <- TCGAanalyze_DEA(mat1 = COAD.expMatrix[,samplesNT],
                            mat2 = COAD.expMatrix[,samplesTP],
                            Cond1type = "Normal",
                            Cond2type = "Tumor",
                            fdr.cut = 0.01 ,
                            logFC.cut = 1,
                            method = "glmLRT")
# Colaprico et al. 2015: fdr.cur = 0.001, logFC.cut >=3

dim(dea.coad)
dea.coad.FiltLevel <- TCGAanalyze_LevelTab(dea.coad,"Tumor","Normal",
                                          COAD.expMatrix[,samplesTP], COAD.expMatrix[,samplesNT])
head(dea.coad.FiltLevel)

starburst.coad_2 <- TCGAvisualize_starburst(met.coad, 
                                     dea.coad,
                                     group1 = "Solid Tissue Normal",
                                     group2 = "Primary solid Tumor",
                                     met.platform = "450K",
                                     genome = "hg19",
                                     filename = "starbustCOAD_sin2.png",
                                     met.p.cut = 10^-2, 
                                     exp.p.cut = 10^-2,
                                     diffmean.cut = 0.25,
                                     logFC.cut = 1,width = 15,height = 10,
                                     names = FALSE)
# Colaprico et al. 2015: logFC.cut >=3

dim(starburst.coad_2)
table(starburst.coad_2['starburst.status'])
```


2. BRCA

```{r starburst.brca}
#dataNorm <-  TCGAanalyze_Normalization(tabDF = exp.brca, geneInfo =  geneInfo, method = 'gcContent') 
#dataNorm2 <-  TCGAanalyze_Normalization(tabDF = exp.brca, geneInfo =  geneInfo, method = 'geneLength')
#dataFilt <- TCGAanalyze_Filtering(tabDF = dataNorm2, method = "quantile", qnt.cut =  0.25)

#par(mfrow=c(2,2))
#boxplot(assay(exp.brca), main = "Datos sin normalizar")
#boxplot(dataNorm, main = "M?todo GC content")
#boxplot(dataNorm2, main = " M?todo geneLength")
#boxplot(dataFilt, main = "Datos normalizads \n + filtrados")


BRCA.expMatrix <- assay(exp.brca)
samplesNT <- which(exp.brca$definition == 'Solid Tissue Normal')
samplesTP <- which(exp.brca$definition == 'Primary solid Tumor')

dea.brca <- TCGAanalyze_DEA(mat1 = BRCA.expMatrix[,samplesNT],
                            mat2 = BRCA.expMatrix[,samplesTP],
                            Cond1type = "Normal",
                            Cond2type = "Tumor",
                            fdr.cut = 0.01 ,
                            logFC.cut = 1,
                            method = "glmLRT")
# Colaprico et al. 2015: fdr.cur = 0.001, logFC.cut >=3

dim(dea.brca)
dea.brca.FiltLevel <- TCGAanalyze_LevelTab(dea.brca,"Tumor","Normal",
                                          BRCA.expMatrix[,samplesTP], BRCA.expMatrix[,samplesNT])
head(dea.brca.FiltLevel)

starburst.brca_2 <- TCGAvisualize_starburst(met.brca, 
                                     dea.brca,
                                     group1 = "Solid Tissue Normal",
                                     group2 = "Primary solid Tumor",
                                     met.platform = "450K",
                                     genome = "hg19",
                                     filename = "starbustBRCA_sin2.png",
                                     met.p.cut = 10^-2, 
                                     exp.p.cut = 10^-2,
                                     diffmean.cut = 0.25,
                                     logFC.cut = 1,width = 15,height = 10,
                                     names = FALSE)
# Colaprico et al. 2015: logFC.cut >=3

# dim(starburst.brca)
table(starburst.brca_2['starburst.status'])
```


3. PRAD

```{r starburst.prad}
PRAD.expMatrix <- assay(exp.prad)
samplesNT <- which(exp.prad$definition == 'Solid Tissue Normal')
samplesTP <- which(exp.prad$definition == 'Primary solid Tumor')

dea.prad <- TCGAanalyze_DEA(mat1 = PRAD.expMatrix[,samplesNT],
                            mat2 = PRAD.expMatrix[,samplesTP],
                            Cond1type = "Normal",
                            Cond2type = "Tumor",
                            fdr.cut = 0.01 ,
                            logFC.cut = 1,
                            method = "glmLRT")
# Colaprico et al. 2015: fdr.cur = 0.001, logFC.cut >=3

dim(dea.prad)
dea.prad.FiltLevel <- TCGAanalyze_LevelTab(dea.prad,"Tumor","Normal",
                                          PRAD.expMatrix[,samplesTP], PRAD.expMatrix[,samplesNT])
head(dea.prad.FiltLevel)

starburst.prad_2 <- TCGAvisualize_starburst(met.prad, 
                                     dea.brca,
                                     group1 = "Solid Tissue Normal",
                                     group2 = "Primary solid Tumor",
                                     met.platform = "450K",
                                     genome = "hg19",
                                     filename = "starbustPRAD_sin2.png",
                                     met.p.cut = 10^-2, 
                                     exp.p.cut = 10^-2,
                                     diffmean.cut = 0.25,
                                     logFC.cut = 1,width = 15,height = 10,
                                     names = FALSE)
# Colaprico et al. 2015: logFC.cut >=3

dim(starburst.prad_2)
table(starburst.prad_2['starburst.status'])
```


4. LUAD
```{r starburst.luad}
LUAD.expMatrix <- assay(exp.luad)
samplesNT <- which(exp.luad$definition == 'Solid Tissue Normal')
samplesTP <- which(exp.luad$definition == 'Primary solid Tumor')

dea.luad <- TCGAanalyze_DEA(mat1 = LUAD.expMatrix[,samplesNT],
                            mat2 = LUAD.expMatrix[,samplesTP],
                            Cond1type = "Normal",
                            Cond2type = "Tumor",
                            fdr.cut = 0.01 ,
                            logFC.cut = 1,
                            method = "glmLRT")
# Colaprico et al. 2015: fdr.cur = 0.001, logFC.cut >=3

dim(dea.luad)
dea.luad.FiltLevel <- TCGAanalyze_LevelTab(dea.luad,"Tumor","Normal",
                                          LUAD.expMatrix[,samplesTP], LUAD.expMatrix[,samplesNT])
head(dea.luad.FiltLevel)

starburst.luad_2 <- TCGAvisualize_starburst(met.luad, 
                                     dea.luad,
                                     group1 = "Solid Tissue Normal",
                                     group2 = "Primary solid Tumor",
                                     met.platform = "450K",
                                     genome = "hg19",
                                     filename = "starbustLUAD_sin2.png",
                                     met.p.cut = 10^-2, 
                                     exp.p.cut = 10^-2,
                                     diffmean.cut = 0.25,
                                     logFC.cut = 1,width = 15,height = 10,
                                     names = FALSE)
# Colaprico et al. 2015: logFC.cut >=3

dim(starburst.luad_2)
table(starburst.luad_2['starburst.status'])
```


## PROBES DIFERENCIALES: 

A partir de los datos preprocesados de metilaci?n `met.coad`, que contienen los valores beta de metilaci?n de cada una de las muestras. En primer lugar se calculan los valores M, m?s adaptados a los c?lculos estad?sticos.

```{r beta2m.func}
##' function to transform the beta values in M values
##' @title 
##' @param beta the beta values
##' @return a dataframe with the m values for all probes
##' @author Lorena Ponce

beta2m <- function(beta) {
    m <- log2(beta/(1 - beta))
    return(m)
}
```


```{r get.mVal}
# Transform the beta values in M values

Mval.coad <- apply(assay(met.coad), c(1,2), beta2m)
Mval.brca <- apply(assay(met.brca), c(1,2), beta2m)
Mval.prad <- apply(assay(met.prad), c(1,2), beta2m)
Mval.luad <- apply(assay(met.luad), c(1,2), beta2m)

# Combine all the dataframes 

Mval1 <- merge(Mval.coad, Mval.brca, by='row.names', all = T)
Mval2 <- merge(Mval.prad, Mval.luad, by='row.names', all = T)
MvalF <- merge(Mval1, Mval2, by = "Row.names", all = T)
MvalF <- na.omit(MvalF)
rownames(MvalF) <- MvalF$Row.names
MvalF <- MvalF[-1]
```

### PROBES DIFERENCIALES CON ABC.RAP:

1. COAD

```{r ttest.abc.coad}
ttest <- ttest_data(assay(met.coad), 1, 30, 31, 60, 1e-2)
nrow(ttest)
ttest2 <- ttest_data(Mval.coad, 1, 30, 31, 60, 1e-2)
nrow(ttest2)
```

2. BRCA

```{r ttest.abc.brca}
ttest.brca <- ttest_data(assay(met.brca), 1, 50, 51, 100, 1e-2)
nrow(ttest.brca)
ttest2.brca <- ttest_data(Mval.brca, 1, 50, 51, 100, 1e-2)
nrow(ttest2.brca)
```

3. PRAD

```{r ttest.abc.prad}
ttest.prad <- ttest_data(assay(met.prad), 1, 50, 51, 100, 1e-2)
nrow(ttest.prad)
ttest2.prad <- ttest_data(Mval.prad, 1, 50, 51, 100, 1e-2)
nrow(ttest2.prad)
```

4. LUAD


```{r ttest.abc.luad}
ttest.luad <- ttest_data(assay(met.luad), 1, 28, 29, 56, 1e-2)
nrow(ttest.luad)
ttest2.luad <- ttest_data(Mval.luad, 1, 28, 29, 56, 1e-2)
#nrow(ttest2.luad)
```



### PROBES DIFERENCIALES CON LIMMA:

```{r limma}
# Use limma without the means 
sampleType <- c(rep("tumorCOAD",30), rep("normalCOAD", 30), rep("tumorBRCA", 50), rep("normalBRCA", 50), rep("tumorPRAD", 50), rep("normalPRAD", 50), rep("tumorLUAD",28), rep("normalLUAD",28))

design <- model.matrix(~ 0+ sampleType)
colnames(design) <-sort(unique(sampleType))

fit <- lmFit(MvalF, design)
#fit2 <-lmFit(MvalF, design2)

cont.matrix <- makeContrasts (
      COADsamples = tumorCOAD-normalCOAD,
      BRCAsamples = tumorBRCA-normalBRCA,
      LUADsamples = tumorLUAD-normalLUAD,
      PRADsamples = tumorPRAD-normalPRAD,
      COAD_BRCA = tumorCOAD-tumorBRCA,
      COAD_LUAD = tumorCOAD-tumorLUAD,
      COAD_PRAD = tumorCOAD-tumorPRAD,
      BRCA_LUAD = tumorBRCA-tumorLUAD,
      BRCA_PRAD = tumorBRCA-tumorPRAD,
      PRAD_LUAD = tumorPRAD-tumorLUAD,
      levels=design)

fit2 <- contrasts.fit(fit, cont.matrix)
efit <- eBayes(fit2)

#ann450kSub <- ann450k[match(rownames(MvalF),ann450k$Name), c(1:4,12:19,24:ncol(ann450k))]
summary(decideTests(efit, adjust.method = "BH", p.value = 0.01))

p.valEfit <- efit$p.value
```
Valores muy altos en comparaci?n con los obtenidos a partir de TCGAanlyze_DMR. A pesar de ello parece que no comparten un mismo perfil de metilaci?n las diferents 

```{r topTableNT}
#ann450kSub <- ann450k[match(rownames(MvalF),ann450k$Name),c(1:4,12:19,24:ncol(ann450k))]
#DMPs <- topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)

ntCOAD <- topTable (efit, number=nrow(efit), coef="COADsamples", adjust="fdr", p.value = 0.01)
ntBRCA <- topTable (efit, number=nrow(efit), coef="BRCAsamples", adjust="fdr", p.value = 0.01)
ntPRAD <- topTable (efit, number=nrow(efit), coef="PRADsamples", adjust="fdr", p.value = 0.01)
ntLUAD <- topTable (efit, number=nrow(efit), coef="LUADsamples", adjust="fdr", p.value = 0.01)
```


```{r}
cat ("En la comparación 'normal vs tumor' en COAD:", sum(ntCOAD$adj.P.Val<=0.01),"\n")
cat ("En la comparación 'normal vs tumor' en BRCA:", sum(ntBRCA$adj.P.Val<=0.01),"\n")
cat ("En la comparación 'normal vs tumor' en PRAD:", sum(ntPRAD$adj.P.Val<=0.01),"\n")
cat ("En la comparación 'normal vs tumor' en LUAD:", sum(ntLUAD$adj.P.Val<=0.01),"\n")
```


```{r volcanoPlot}
for(i in 1:4){
  compName <-colnames(cont.matrix)[i]
  file=paste("volcanoPlot", compName, ".pdf", sep="")
  pdf(file=file.path("E:/Escritorio/Epigenetics & cancer", file), paper="special", width=6, height=6)
  volcanoplot(efit, coef=i, highlight=10, 
            main=paste("Differentially methylated probes",compName, sep="\n"))
  lp <- -log10(efit$p.value[,i])
  ord <- order(lp, decreasing = TRUE)[1:10]
  points(fit2$coef[ord,i], lp[ord], pch = 16, cex = 0.45, col = "blue")
  abline(v=c(-1,1))
  dev.off()
  cat("\\includegraphics{", file, "}\n\n", sep="")
}
```


#### Comparación de los 3 paquetes de análisis
A continuación se comparan el número de probes significativas en cada uno de los métodos empleados:
En primer lugar para COAD: 

```{r check.sig.probes.coad}
sig.coad <- ntCOAD[which(ntCOAD$adj.P.Val <= 0.01),]
sig2 <- p.valEfit[which(p.valEfit[,1]<= 0.01),]
length(which(rownames(sig.coad) %in% rownames(sig2)))
length(which(rownames(sig.coad) %in% dmr_coad_sel$probeID))
length(which(rownames(sig.coad) %in% rownames(ttest2)))
length(which(dmr_coad_sel$probeID %in% rownames(ttest2)))
```

En segundo lugar para BRCA:

```{r check.sig.probes.brca}
sig.brca <- ntBRCA[which(ntBRCA$adj.P.Val <= 0.01),]
sig2 <- p.valEfit[which(p.valEfit[,2]<= 0.01),]
length(which(rownames(sig.brca) %in% rownames(sig2)))
length(which(rownames(sig.brca) %in% dmr_brca_sel$probeID))
length(which(rownames(sig.brca) %in% rownames(ttest2.brca)))
length(which(dmr_brca_sel$probeID %in% rownames(ttest2.brca)))
```

En tercer lugar para PRAD:

```{r check.probes.sig.prad}
sig.prad <- ntPRAD[which(ntPRAD$adj.P.Val <= 0.01),]
sig2 <- p.valEfit[which(p.valEfit[,4]<= 0.01),]
length(which(rownames(sig.prad) %in% rownames(sig2)))
length(which(rownames(sig.prad) %in% dmr_prad_sel$probeID))
length(which(rownames(sig.prad) %in% rownames(ttest2.prad)))
length(which(dmr_prad_sel$probeID %in% rownames(ttest2.prad)))
```

Y finalmente para LUAD, pero en este caso no se pudo realizar el ttest del paquete ABC.RAP con los valores M.

```{r check.sig.probes.luad}
sig.luad <- ntLUAD[which(ntLUAD$adj.P.Val <= 0.01),]
sig2 <- p.valEfit[which(p.valEfit[,3]<= 0.01),]
length(which(rownames(sig.luad) %in% rownames(sig2)))
length(which(rownames(sig.luad) %in% dmr_luad_sel$probeID))
length(which(rownames(sig.luad) %in% rownames(ttest.luad)))
length(which(dmr_luad_sel$probeID %in% rownames(ttest.luad)))
```


#### Extracción de las probes significativas y cálculo de los valores beta

```{r}
sigCOAD <- assay(met.coad)[which(row.names(assay(met.coad)) %in% row.names(ntCOAD)),]
meansTumor <- rowMeans(sigCOAD[,grep("01A", colnames(sigCOAD))], na.rm=TRUE)
meansNormal <- rowMeans(sigCOAD[,grep("11A", colnames(sigCOAD))], na.rm=TRUE)
means <- data.frame("meansTumor" = meansTumor, "meansNormal" = meansNormal)
means$dif <- (means$meansTumor - means$meansNormal)
length(which(abs(means$dif) >= 0.7))
length(which(abs(means$dif) <= 0.2))
```


```{r}
#m3 <-as.data.frame(merge(ann450kSub, ntCOAD, by.x = "Name", by.y = "row.names", all.x = FALSE, all.y = TRUE))
```


#### Clasificación de los perfiles de metilación


```{r}
# extracción de los valores beta de las muestras tumorales y de las probes significativas
sigCOADt <- t(sigCOAD[,grep("01A", colnames(sigCOAD))])
```

```{r}
library(cluster)
a <- agnes(sigCOADt, method="average")
pdf(file = "aAverage.pdf", width = 17, height = 12, family = "Helvetica")
plot(a, which.plots = 2, main = "Dendogram with Average method")
#rect.hclust(agnes, k=4, border="red")
#dev.off()
```

```{r}
# Elbow method
fviz_nbclust(sigCOADt, kmeans, method = "wss") +
    geom_vline(xintercept = 4, linetype = 2)+
  labs(subtitle = "Elbow method")

# Silhouette method
fviz_nbclust(sigCOADt, kmeans, method = "silhouette")+
  labs(subtitle = "Silhouette method")

# Gap statistic
# nboot = 50 to keep the function speedy. 
# recommended value: nboot= 500 for your analysis.
# Use verbose = FALSE to hide computing progression.
set.seed(123)
fviz_nbclust(sigCOADt, kmeans, nstart = 25,  method = "gap_stat", nboot = 50)+
  labs(subtitle = "Gap statistic method")
```

```{r}
library(NbClust)
NbClust(sigCOADt, distance = "euclidean", min.nc = 2, max.nc = 8, method = "kmeans")
```


```{r}
set.seed(12345)
# Formamos 4 grupos por k-means
km <- kmeans(sigCOADt, centers=4, algorithm="Hartigan-Wong")
km$size
km$cluster
#pred_table <- table(kmeans=km$cluster); pred_table
```





## COMPARACIÓN DE LOS PERFILES DE METILACÓN ENTRE COHORTES

```{r topTable.cohorts, eval=FALSE}
#ann450kSub <- ann450k[match(rownames(MvalF),ann450k$Name),c(1:4,12:19,24:ncol(ann450k))]
#DMPs <- topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)

COADvsBRCA <- topTable (efit, number=nrow(efit), coef="COAD_BRCA", adjust="fdr")
COADvsPRAD <- topTable (efit, number=nrow(efit), coef="COAD_PRAD", adjust="fdr")
COADvsLUAD <- topTable (efit, number=nrow(efit), coef="COAD_LUAD", adjust="fdr")
BRCAvsLUAD <- topTable (efit, number=nrow(efit), coef="BRCA_LUAD", adjust="fdr")
BRCAvsPRAD <- topTable (efit, number=nrow(efit), coef="BRCA_PRAD", adjust="fdr")
PRADvsLUAD <- topTable (efit, number=nrow(efit), coef="PRAD_LUAD", adjust="fdr")
```



```{r rowMeans.betaVal}
# Get row means of M values 

rowMean.coad <- data.frame(normalMeans.COAD = (rowMeans(Mval.coad[,31:60])), tumorMeans.COAD =(rowMeans(Mval.coad[,1:30])),  row.names = row.names(Mval.coad))
rowMean.brca <- data.frame(normalMeans.BRCA = (rowMeans(Mval.brca[,51:100])), tumorMeans.BRCA =(rowMeans(Mval.brca[,1:50])),  row.names = row.names(Mval.brca))
rowMean.prad <- data.frame(normalMeans.PRAD = (rowMeans(Mval.prad[,51:100])), tumorMeans.PRAD =(rowMeans(Mval.prad[,1:50])),  row.names = row.names(Mval.prad))
rowMean.luad <- data.frame(normalMeans.LUAD = (rowMeans(Mval.luad[,29:56])), tumorMeans.LUAD =(rowMeans(Mval.luad[,1:28])),  row.names = row.names(Mval.luad))


# Combine the rowMean of all cohorts

rowMean_Mval <- merge(rowMean.brca, rowMean.coad, by='row.names', all = T)
rowMean_Mval2 <- merge(rowMean.luad, rowMean.prad, by='row.names', all = T)
Mval <- merge(rowMean_Mval, rowMean_Mval2, by = "Row.names", all = T)
dim(Mval)
Mval_t <- na.omit(Mval)
rownames(Mval_t) <- Mval_t$Row.names
Mval_t <- Mval_t[-1]
dim(Mval_t)
#cpg <- row.names(Mval_t)
```


En primer lugar, para comparar las diferencias entre los distintos tipos de tumores vamos a comprobar la representación de las medias en un gráfico MDS:

```{r plotMDS}
plotMDS(Mval_t, top=1000, gene.selection="common", col=brewer.pal(ncol(Mval_t), "Spectral"))
abline(v=0, h=0, col="blue")
legend("top", legend=colnames(Mval_t), text.col=brewer.pal(ncol(Mval_t), "Spectral"), bg="white", cex=0.7)
dev.print(pdf, 'MDSmean.pdf')
```

Las componentes principales parecen indicar que las medias de los tumores se encuentran, en general, más cercanos a los valores positivos de la segunda componente principal (excepto tumor LUAD), mientras que las muestras normales se encontrarían más abajo en el gráfico (excepto normal PRAD). Por otra parte, parece que la primera componete principal divide las muestras de COAD con el resto de tipos de tumores, a excepción de las muestras tumorales de LUAD que se sitúan cercanas a 0. 

Al volver a comprobar los valores obtenidos con limma vemos que, desde el mismo punto de partida, hay menos diferencias significativas al comparar COAD-BRCA y BRCA-PRAD, seguido de COAD-PRAD. En este caso únicamente se compararan las muestras tumorales, para comprobar si presentan los mismos o diferentes perfiles de metilación. Por lo qeu coincidiría con el gráfico MDS, donde las medias del tejido tumoral localizadas en los valores positivos de la segunda componente principal presentan menos prodes significativamente diferentes. Por el contrario, las que más diferencias significativas presentan son las comparadas con LUAD. 

```{r}
summary(decideTests(efit, adjust.method = "BH", p.value = 0.01))
```


```{r transposeMval}
#sig.probes <- unique(c(row.names(sig.coad),row.names(sig.brca), row.names(sig.luad), row.names(sig.prad)))
class <- as.data.frame(c(rep("tumor-COAD", 30), rep("tumor-BRCA", 49), rep("tumor-PRAD", 50),  rep("tumor-LUAD", 25)))

MvalF2 <- MvalF[is.finite(rowSums(MvalF)),]
MvalF3 <- MvalF2[, grep("01A", colnames(MvalF2))]
#MvalF2 <- MvalF2[which(row.names(MvalF2) %in% sig.probes),]
#MvalF2 <- MvalF2[sig.probes,]
#head(which(row.names(MvalF2) %in% sig.probes))
#MvalF2 <- MvalF2[-c(4,5,8,9,10),]

Mvalt <- t(MvalF3)
Mvalt <- cbind(Mvalt, class)
colnames(Mvalt)[368647]<- "class"
disE = dist(Mvalt[,-368647])
```

El cálculo MDs a partir de la distancias Euclídeas de todas las muestras tumorales se muestra a continuación:

```{r calculeMDS}
mds <- cmdscale(disE,eig=T)
#jpeg(file="MDSdisE.jpeg", res = 300)
pdf(file = "MDSdisE.pdf", width = 17, height = 12, family = "Helvetica")
plot(mds$points,xlab="Coord 1",ylab="Coord 2", col=as.numeric(Mvalt[,368647])+1)
abline(h=0,v=0,lty=4,col="gray")
title("MDS with Euclidean distance")
legend("topleft", legend=levels(Mvalt[,368647]), pch=21, col = 2:5)
dev.off()
```

A pesar de que la distribución es diferente a la observada comparando las medias, BRCA y PRAD se sitúan a la derecha de la primera coordenada y LUAD y COAD a la izquierda, de manera similar a lo qeu ocurría con la primera componente de la PCA. 

```{r accumPerc}
evv <- mds$eig
pb <- cumsum(evv)/sum(evv)*100; round(pb[1:6],2)
names(pb) <- paste("Coord",1:6,sep = "")
pdf(file = "AccPerc.pdf", width = 17, height = 12, family = "Helvetica")
barplot(pb[1:6],ylab="porcentaje")
title(main="Accumulated percentage",line=1.5)
dev.off()
```

A pesar de esta relación, y debido al elevado número de probes y a la variabilidad que estos presentan, esta tendencia únicamente explica cerca del 20% de la variabilidad encontrada; y el conjunto de ambas coordenadas no superan el 30% de la variabilidad explicada. A pesar de ello es posible encontrar similitudes con el gráfico MDS calculado a partir de las medias. 


```{r lattice}
pdf(file = "lattice.pdf", width = 17, height = 12, family = "Helvetica")
lattice::levelplot(as.matrix(disE))
dev.off()
```


Seguidamente se clasificarán las muestras mediante el empleo de dendogramas o árboles de clasificación: 

```{r dendogramaCOMPLETE}
# MÉTODO COMPLETE 

hccomp <- hclust(disE, method="complete")
pdf(file = "hccomp.pdf", width = 17, height = 12, family = "Helvetica")
plot(hccomp, labels = Mvalt$class, main= "Dendogram with complete method")
rect.hclust(hccomp, k=4, border="red")
dev.off()
```

```{r resCOMPLETE}
hccomp_res <- cutree(hccomp,4)
table(Cluster=hccomp_res ,Tipo=Mvalt$class)
```

Con el método complete la primera clase o cluster está más relacionada con COAD, la segunda con prácticamente todos los tipos de tumor, mientras qeu la 3a y 4a apenas tienen correspondencia con el tipo de tumor. 

A continuación se empleará el método de Ward:

```{r dendogramaWARD}
# MÉTODO DE WARD
agnes <- agnes(Mvalt[,-368647], method="ward")
pdf(file = "agnes.pdf", width = 17, height = 12, family = "Helvetica")
plot(agnes, which.plots = 2, labels = Mvalt$class, main = "Dendogram with Ward method")
rect.hclust(agnes, k=4, border="red")
dev.off()
```

```{r resWARD}
agnes_res <- cutree(agnes, k=4)
table(Ward=agnes_res, Type=Mvalt$class)
```

A partir del método de Ward se clasifican correctamente el 100% de las muestras, donde el primer cluster se relaciona con COAD, el segudo con BRCA, el tercero con PRAD y el cuarto con LAUD. Al observar el dondegrama de clasficación se aprecia el mismo patrón que el observado en los gráficos MDS: LUAD y COAD están más próximos entre sí, al igual que BRCA y PRAD.

Seguidamente se procede a realizar un partición mediante la técnica de k-medias o k-means. Previo a su realización comprobamos el número de k apropiado para hacer la clasificación:

```{r nK}
set.seed(123)
wss <- rep(0, 8) 
for(i in 1:8){wss[i] <- sum(kmeans(Mvalt[,-368647], centers= i)$withinss)}
pdf(file = "kSel.pdf", width = 17, height = 12, family = "Helvetica")
plot(1:8, wss, type = "b", xlab = "Number of groups", ylab = "Within groups sum of squares")
dev.off()
```

```{r K-means}
set.seed(12345)
# Formamos 4 grupos por k-means
km <- kmeans(Mvalt[,-368647], centers=4, algorithm="Hartigan-Wong"); km
km$size
pred_table <- table(kmeans=km$cluster, Type=Mvalt$class); pred_table
```

```{r fviz}
pdf(file = "kmCluster.pdf", width = 17, height = 12, family = "Helvetica")
fviz_cluster(km, Mvalt[, -368647], legend = "right", geom = "points", pointsize = 12, ggtheme = theme_minimal())
dev.off()
```


```{r pca}
pca <- prcomp(Mvalt[,-368647], scale=TRUE)
#summary(pca)
```


Seguidamente se va a comprobar la homogeneidad de varianzas para comprobar si sería útil un análisis discriminante:

```{r}
p <- ncol(Mvalt[,-368647])
n <- table(Mvalt[368647])
S1 <- (n[1]-1)*cov(Mvalt[Mvalt$class=="tumor-BRCA",-368647])/n[1]
S2 <- (n[2]-1)*cov(Mvalt[Mvalt$class=="tumor-COAD",-368647])/n[2]
S3 <- (n[3]-1)*cov(Mvalt[Mvalt$class=="tumor-LUAD",-368647])/n[3]
S4 <- (n[4]-1)*cov(Mvalt[Mvalt$class=="tumor-PRAD",-368647])/n[4]
S <- (n[1]*S1+n[2]*S2+n[3]*S3+n[4]*S4)/sum(n)
llr <- sum(n)*log(det(S))-n[1]*log(det(S1))-n[2]*log(det(S2))-n[3]*log(det(S3))-n[4]*log(det(S4))
p.valor <- pchisq(llr, p*(p+1)/2, lower.tail=FALSE)
c(llr=as.numeric(llr),p.valor=as.numeric(p.valor))
```


Finalmente se emplea la función `sknn()` del paquete `klaR` para estudiar la distancia con un modelo simple de vecinos cercanos, a partir de la matriz de distancias Euclídeas, que como vimos anteriormente clasifica adecuadamente los datos. 

```{r}
knn.res <- sknn(class ~ ., data = Mvalt)
table(Type=predict(knn.res)$class,Mvalt$class)
training_error <- mean(Mvalt$class != predict(knn.res)$class) *100; training_error
```












# AN?LISIS MUTACIONAL

```{r}
mut.coad <- as.data.frame(assay(maf.coad))
mut.brca <- as.data.frame(assay(maf.brca))
mut.luad <- as.data.frame(assay(maf.luad))
mut.prad <- as.data.frame(assay(maf.prad))

mut.coad <- mut.coad[,strtrim(colnames(mut.coad), 12)%in% strtrim(barcodes.coad, 12)] #no coincide ninguna columna
dim(mut.coad)

mut.brca <- mut.brca[,strtrim(colnames(mut.brca), 12)%in% strtrim(barcodes.brca, 12)]
dim(mut.brca)

mut.luad <- mut.luad[,strtrim(colnames(mut.luad), 12)%in% strtrim(barcodes.luad, 12)] 
dim(mut.luad)

mut.prad <- mut.prad[,strtrim(colnames(mut.prad), 12)%in% strtrim(barcodes.prad, 12)] 
dim(mut.prad)
```


## CÓDIGO NO USADO

```{r eval=FALSE}
##NOT USED!!
# DMR with limma

#my.targets <-read.AnnotatedDataFrame("targets.txt", header = TRUE) 
sampleType <- colnames(Mval_t)
tumorType <- c(rep("BRCA", 2), rep("COAD", 2), rep("LUAD", 2), rep("PRAD", 2))
tissueType <- c(rep(c("normal", "tumor"), 4))

design <- model.matrix(~0+sampleType)
colnames(design) <- sampleType
print(design)

fit <- lmFit(Mval_t, design)

cont.matrix <- makeContrasts (
      COADsamples = tumorMeans.COAD-normalMeans.COAD,
      BRCAsamples = tumorMeans.BRCA-normalMeans.BRCA,
      LUADsamples = tumorMeans.LUAD-normalMeans.LUAD,
      PRADsamples = tumorMeans.PRAD-normalMeans.PRAD,
      COAD_BRCA = tumorMeans.COAD-tumorMeans.BRCA,
      COAD_LUAD = tumorMeans.COAD-tumorMeans.LUAD,
      COAD_PRAD = tumorMeans.COAD-tumorMeans.PRAD,
      BRCA_LUAD = tumorMeans.BRCA-tumorMeans.LUAD,
      BRCA_PRAD = tumorMeans.BRCA-tumorMeans.PRAD,
      PRAD_LUAD = tumorMeans.PRAD-tumorMeans.LUAD,
      levels=design)

fit2 <- contrasts.fit(fit, cont.matrix)
efit <- eBayes(fit2)
```



```{r eval=FALSE}
# Get the difference between tumor and normal and save it in a data.frame

z <- unique(strtrim(colnames(Mval.coad), 12))

deltaVal <- data.frame(matrix(nrow = nrow(Mval.coad), ncol = ncol(Mval.coad)))
colnames(deltaVal) <- z

for (row in 1:nrow(Mval.coad)){
  for (i in 1:length(z)) {
    x <- which(strtrim(colnames(Mval.coad), 12) == z[i])
    deltaVal[row,i] <- round(Mval.coad[row, x[2]] - Mval.coad[row, x[1]], 2)
    if (i == 30 & row == nrow(Mval.coad)) cat("Done")
    progress(i, row)
  }
}
```


# Save image

```{r}
#save.image(file="myObjects.RData")
```



# Session

```{r sessionInfo, cache = FALSE}
devtools::session_info()
```

